In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
# allows the notebook to render graphics
%matplotlib inline
# a popular data visualization theme
plt.style.use('fivethirtyeight')

In [2]:
# import first dataset to explore salaries in San Francisco
# https://data.sfgov.org/City-Management-and-Ethics/Employee-Compensation/88g8-5mnd

salary_ranges = pd.read_csv("./data/Employee_Compensation.csv")



/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [9]:
salary_ranges = pd.read_json("https://data.sfgov.org/resource/88g8-5mnd.json")

In [10]:
# looking at table to get a better understanding of the data
salary_ranges.head()


department_code  employee_identifier  health_and_dental  \
0             CRT              8540990                  0   
1             CRT              8540990                  0   
2             HSA              8530592                  0   
3             DPH              8499266                  0   
4             HSA              8571272                  0   

                            job job_code           job_family job_family_code  \
0         Deputy Court Clerk II     420C             Untitled            0000   
1         Deputy Court Clerk II     420C             Untitled            0000   
2  Public Svc Aide-Public Works     9916  Public Service Aide            9900   
3   Senior Physician Specialist     2232     Medical & Dental            2200   
4        Public Service Trainee     9910  Public Service Aide            9900   

                         organization_group  organization_group_code  \
0                         Public Protection                        1   
1                         Public Protection                        1   
2  Human Welfare & Neighborhood Development                        3   
3                          Community Health                        4   
4  Human Welfare & Neighborhood Development                        3   

   other_benefits    ...      overtime  retirement  salaries  total_benefits  \
0           53.86    ...          0.00      130.91    674.28          184.77   
1           53.86    ...          0.00      130.91    674.28          184.77   
2          518.10    ...        578.58        0.00   5723.82          518.10   
3         5552.57    ...          0.00    15724.50  38319.67        21277.07   
4           76.83    ...          0.00        0.00      0.00           76.83   

   total_compensation  total_salary                           union  \
0              864.81        680.04  Utd Pub EmpL790 SEIU-Crt Clrks   
1              864.81        680.04  Utd Pub EmpL790 SEIU-Crt Clrks   
2             6991.27       6473.17             Laborers, Local 261   
3            74070.33      52793.26   Physician/Dentists 8-CC, UAPD   
4             1081.26       1004.43          SEIU, Local 1021, Misc   

  union_code  year  year_type  
0        792  2028   Calendar  
1        792  2028     Fiscal  
2        261  2019   Calendar  
3        164  2019   Calendar  
4        790  2019   Calendar  

[5 rows x 21 columns]

In [11]:
salary_ranges.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 21 columns):
department_code            1000 non-null object
employee_identifier        1000 non-null int64
health_and_dental          1000 non-null int64
job                        1000 non-null object
job_code                   1000 non-null object
job_family                 1000 non-null object
job_family_code            1000 non-null object
organization_group         1000 non-null object
organization_group_code    1000 non-null int64
other_benefits             1000 non-null float64
other_salaries             1000 non-null float64
overtime                   1000 non-null float64
retirement                 1000 non-null float64
salaries                   1000 non-null float64
total_benefits             1000 non-null float64
total_compensation         1000 non-null float64
total_salary               1000 non-null float64
union                      1000 non-null object
union_code               

We can see there are approximately 700,000 entries (rows) and 22 columns. The *.info()* command is important because it tells the number of non-null items in each column, and issing data is one of the most common issues in feature engineering. There are many ways to deal with missing values, and one quick way to count the number of missing values is to run *.isnull().sum()* on the dataframe.

In [7]:
salary_ranges.isnull().sum()

Year Type                       0
Year                            0
Organization Group Code         0
Organization Group              0
Department Code                 0
Department                 273993
Union Code                    514
Union                         514
Job Family Code                 0
Job Family                      0
Job Code                        0
Job                             2
Employee Identifier             0
Salaries                        0
Overtime                        0
Other Salaries                  0
Total Salary                    0
Retirement                      0
Health and Dental               0
Other Benefits                  0
Total Benefits                  0
Total Compensation              0
dtype: int64

Note: There are four columns with missing values. I am going to drop the two missing records in the *Job* column, and I will remove column *Union Code* from the dataset, but I will leave the other two columns with missing data alone (since I will not be using this data).

In [20]:
# drop 2 records with missing data in Job column
salary_ranges.dropna(subset=['Job'], inplace = True)

In [21]:
len(salary_ranges)

697228

As we can see there were originally 697230 entries, there are now only 697228.

In [26]:
# remove union code from data set
salary_ranges = salary_ranges.drop(["Union Code"], axis = 1)

In [27]:
for i in salary_ranges.columns:
    print(i)

Year Type
Year
Organization Group Code
Organization Group
Department Code
Department
Union
Job Family Code
Job Family
Job Code
Job
Employee Identifier
Salaries
Overtime
Other Salaries
Total Salary
Retirement
Health and Dental
Other Benefits
Total Benefits
Total Compensation


As we can see *Union Code* is no longer in the dataset.

Now let's run *describe* method to look at some descriptive statistics of the quantitative columns. One thing to note, the *describe* method will default to describing quantitative columsn only if they exist, otherwise it will describe qualitative columns.

In [28]:
# show describptive stats
salary_ranges.describe()

Year  Organization Group Code  Employee Identifier  \
count  697228.000000            697228.000000         6.972280e+05   
mean     2016.228136                 3.312343         3.384903e+06   
std         1.886411                 1.882790         4.172615e+06   
min      2013.000000                 1.000000         1.000000e+00   
25%      2015.000000                 2.000000         2.345600e+04   
50%      2017.000000                 3.000000         4.703900e+04   
75%      2018.000000                 5.000000         8.533240e+06   
max      2028.000000                 7.000000         1.041014e+07   

            Salaries       Overtime  Other Salaries   Total Salary  \
count  697228.000000  697228.000000   697228.000000  697228.000000   
mean    52856.756051    3971.379239     2861.740261   59689.875551   
std     45831.587739   10685.974994     6502.309301   53055.830916   
min    -68771.780000  -18458.150000   -19131.100000  -68771.780000   
25%      7214.197500       0.000000        0.000000    8034.375000   
50%     51450.410000       0.000000      250.000000   55908.510000   
75%     82627.045000    1965.852500     2852.520000   92570.402500   
max    616964.710000  309897.200000   342802.630000  622102.650000   

          Retirement  Health and Dental  Other Benefits  Total Benefits  \
count  697228.000000      697228.000000   697228.000000   697228.000000   
mean    10390.778883        7491.307354     3944.018028    21761.637942   
std      9586.763455        5368.582115     3921.857444    17550.768163   
min    -30621.430000       -2940.470000   -10636.500000   -21295.150000   
25%         0.000000        1955.735000      529.637500     2185.757500   
50%     10282.025000        7566.940000     3136.910000    24023.240000   
75%     16880.055000       12791.455000     6225.992500    35704.280000   
max    115554.870000       22270.120000    37198.600000   141043.640000   

       Total Compensation  
count       697228.000000  
mean         81451.513493  
std          69605.441996  
min         -74082.610000  
25%          10817.135000  
50%          80865.900000  
75%         128097.320000  
max         759312.180000

Let's ignore most of these columns and focus in on *Salaries*. Note, although the average salary is approximately $50,000; I question this accuracy due to the minimum salary being a negative number (i.e. -68771.78). Further investigation would need to be done on this series. This will have to be covered in another section. Moving forward let's simply look at the first few rows of job title and salaries.

In [31]:
salary_ranges[['Job','Salaries']].head(10)

Job  Salaries
0           Deputy Court Clerk II    674.28
1           Deputy Court Clerk II    674.28
2  StdntDsgnTrain1, Arch/Eng/Plng      0.00
3   Recreation Facility Assistant    561.60
4    Lieutenant, Fire Suppression  87390.00
5                   Special Nurse  25369.90
6               Recreation Leader   1913.64
7                   Special Nurse   5147.10
8       Pr Administrative Analyst  79865.75
9                  Pool Lifeguard   7994.25

Here we can see the first 10 rows. I found it interesting the the third row has a salary of zero. I am curious if this number is accurate, or if it reflects an unpaid role (i.e. intern, or student helper).